proposals/nnnn-placeholder-types.md

```diff
@@ -0,0 +1,349 @@
+
+### Generic parameter inference
+
+A limited version of this feature is already present in the language via generic parameter inference. When the generic arguments to a generic type can be inferred from context, you are permitted to omit them, like so:
+
+```swift
+import Combine
+
+let publisher = Just(0) // Just<Int> is inferred!
+```
+
+With placeholder types, writing the bare name of a generic type (in most cases, see note below) becomes equivalent to writing the generic signature with placeholder types for the generic arguments. E.g., the initialization of `publisher` above is the same as:
+
+```swift
+let publisher = Just<_>(0)
+```
+
+Note: there is an existing rule that *inside the body* of a generic type `S<T1, ..., Tn>` , the bare name `S` is equivalent to `S<T1, ..., Tn>` . This proposal does not augment this rule nor attempt to express this rule in terms of placeholder types.
+
+### Function signatures
+
+As is the case today, function signatures under this proposal are required to have their argument and return types fully specified. Generic parameters cannot be inferred and placeholder types are not permitted to appear within the signature, even if the type could ostensibly be inferred from e.g., a protocol requirement or default argument expression.
+
+Thus, it is an error under this proposal to write something like:
+
+```swift
+func doSomething(_ count: _? = 0) { ... }
+```
+
+just as it would be an error to write:
+
+```swift
+func doSomething(_ count: Optional = 0) { ... }
+```
+
+even though the type checker could infer the `Wrapped` type in an expression like:
+
+```swift
+let count: _? = 0
+```
+
+As a more comprehensive example, consider the following setup:
+
+```swift
+struct Bar<T, U>
+where T: ExpressibleByIntegerLiteral, U: ExpressibleByIntegerLiteral {
+    var t: T
+    var u: U
+}
+
+extension Bar {
+    func frobnicate() -> Bar {
+        return Bar(t: 42, u: 42)
+    }
+    func frobnicate2() -> Bar<_, _> {
+        return Bar(t: 42, u: 42)
+    }
+    func frobnicate3() -> Bar {
+        return Bar<_, _>(t: 42, u: 42)
+    }
+    func frobnicate4() -> Bar<_, _> {
+        return Bar<_, _>(t: 42, u: 42)
+    }
+    func frobnicate5() -> Bar<_, U> {
+        return Bar(t: 42, u: 42)
+    }
+    func frobnicate6() -> Bar {
+        return Bar<_, U>(t: 42, u: 42)
+    }
+    func frobnicate7() -> Bar<_, _> {
+        return Bar<_, U>(t: 42, u: 42)
+    }
+    func frobnicate8() -> Bar<_, U> {
+        return Bar<_, _>(t: 42, u: 42)
+    }
+}
+```
+
+Under this proposal, only `frobnicate1`, `frobnicate3` and `frobnicate6` would compile without error (`frobnicate1`, of course, compiles without this proposal as well), since all others have placeholders appearing in at least one position in the function signature.
```

Comment:
Nit: `frobnicate1` doesn't exist, it's just `frobnicate`.

Reply:
```suggestion
Under this proposal, only `frobnicate`, `frobnicate3` and `frobnicate6` would compile without error (`frobnicate1`, of course, compiles without this proposal as well), since all others have placeholders appearing in at least one position in the function signature.
```